### PR TITLE
Make audio streams handle work as in cleo4

### PR DESCRIPTION
**Pull request by MiranDMC**
Due to adding virtual functions to the audio stream class VMT pointer was added to the data layout. This caused stream handle (address) no longer pointing on property containing underlying BASS stream pointer.
Updated so more advanced scripts still can work with BASS as in CLEO4.

Unit tests working, audio demo script works too.

### CODE REVIEW

File: cleo_plugins/Audio/StreamHandle.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+#include "CLEO_Utils.h"
+#include "CAudioStream.h"
+
+using namespace CLEO;
+
+// scripts receive handles that are not exact pointers to the CAudioStream
+// instead adjusted address pointing "streamInternal" property is used, to mimick as it was in CLEO4
+class StreamHandle
+{
+    static const DWORD Handle_Offset = 4; // skip Virtual Method Table
```
Comment (x87) [inline cleo_plugins/Audio/StreamHandle.h:11]: how can you guarantee that this offset is always 4?

Comment (MiranDMC) [reply cleo_plugins/Audio/StreamHandle.h]: `offsetof` macro was returning 8 instead of 4. After investigation it turned out CLEO4 already had virtual methods in audio stream classes. Additional 4 bytes offset was caused by inheriting from VInjectible, which has VMT too.

File: source/CleoBase.cpp [hunk 2]
```diff
@@ -252,9 +252,9 @@ namespace CLEO
         GetInstance().RemoveCallback(id, func);
     }
 
-    DWORD WINAPI CLEO_GetInternalAudioStream(CLEO::CRunningScript* thread, DWORD stream) // arg CAudioStream *
+    DWORD WINAPI CLEO_GetInternalAudioStream(CLEO::CRunningScript* unused, DWORD audioStreamPtr)
     {
-        return stream; // CAudioStream::streamInternal offset is 0
+        return audioStreamPtr + 0x8; // offsetof(CAudioStream, streamInternal)
```
Comment (x87) [inline source/CleoBase.cpp:257]: why can't you do `audioStreamPtr->GetInternal()` here, like CLEO4?

Comment (MiranDMC) [reply source/CleoBase.cpp:257]: Audio stuff is now separated plugin. I do not think it is worth to create project dependency in that direction to support just this single export.

Comment (MiranDMC) [reply source/CleoBase.cpp:257]: Offset of internal BASS stream in relation to audio stream handle returned by CLEO is untouchable, as some scripts are using it already, just by adding offset and reading memory.
The export was apparently added by Junior for purposes of CLEO+ opcodes.

Comment (x87) [reply source/CleoBase.cpp:257]: so why do you need this class Dummy then? what offset the scripts use?

File: cleo_plugins/Audio/CAudioStream.h [hunk 3]
```diff
@@ -5,7 +5,14 @@
 
 namespace CLEO
 {
-    class CAudioStream
+    // in CLEO4 CAudioStream was extending VInjectible
```
Comment (x87) [inline cleo_plugins/Audio/CAudioStream.h:8]: No, it wasn't

https://github.com/cleolibrary/CLEO4/blob/master/source/CSoundSystem.h#L57

Comment (MiranDMC) [reply cleo_plugins/Audio/CAudioStream.h:8]: Hm, right. 
Then why in CLEO4 offset was 8 and now without adjustments it will be just 4.

Comment (x87) [reply cleo_plugins/Audio/CAudioStream.h:8]: Where did you find offset 8? 

```
.text:100080C0 ; __stdcall CLEO_GetInternalAudioStream(x, x)
.text:100080C0                 public _CLEO_GetInternalAudioStream@8
.text:100080C0 _CLEO_GetInternalAudioStream@8 proc near
.text:100080C0                                         ; DATA XREF: .rdata:off_1004A6B8↓o
.text:100080C0
.text:100080C0 arg_4           = dword ptr  0Ch
.text:100080C0
.text:100080C0                 push    ebp
.text:100080C1                 mov     ebp, esp
.text:100080C3                 mov     eax, [ebp+arg_4]
.text:100080C6                 mov     eax, [eax+4] <------------------------ +4
.text:100080C9                 pop     ebp
.text:100080CA                 retn    8
.text:100080CA _CLEO_GetInternalAudioStream@8 endp
```

Comment (MiranDMC) [reply cleo_plugins/Audio/CAudioStream.h:8]: Ok I had to write my own test script:
```
{$CLEO .cs}
nop

int lib 
if
    not lib = load_dynamic_library {fileName} "bass.dll"
then
    terminate_this_script
end

int proc
if
    not proc = get_dynamic_library_procedure {procName} "BASS_ChannelGetLevel" {library} lib
then
    terminate_this_script
end

int stream = load_audio_stream {audioFileName} "cleo\cleo_tests\Audio\Ding.mp3"
//int stream = load_3d_audio_stream {audioFileName} "cleo\cleo_tests\Audio\Ding.mp3"

set_audio_stream_state stream {action} AudioStreamAction.Play
set_audio_stream_looped stream {state} true


while true
    wait 0
    
    int bassStream = stream + 4
    bassStream = read_memory {address} bassStream {size} 4 {vp} false
    
    int levels = call_function_return {address} proc {numParams} 1 {pop} 0 {funcParams} bassStream

    levels &= 0xFFFF // left channel

    float value =# levels
    value /= 32768.0 // 0xFFFF

    value *= 200.0
    
    float y = value
    y /= -2.0
    y += 400.0
    
    use_text_commands {state} true
    draw_rect {x} 0.0 {y} y {width} 50.0 {height} value {rgba} 128 255 128 255
end

terminate_this_custom_script
```

Comment (x87) [reply cleo_plugins/Audio/CAudioStream.h:8]: try to call `CLEO_GetInternalAudioStream(_, stream)` and compare its result with `bassStream`. I think you're also missing reading the value from the pointer you got. You should return the stream value, not the address

Comment (MiranDMC) [reply cleo_plugins/Audio/CAudioStream.h:8]: This script works in CLEO4. For some reason in CLEO5 it requires
```
int bassStream = stream + 8
```
to work.
I think it should be de referenced once, as this is Bass stream handle value.

Comment (x87) [reply cleo_plugins/Audio/CAudioStream.h:8]: did you remove the extra dummy class?

Comment (MiranDMC) [reply cleo_plugins/Audio/CAudioStream.h:8]: I added compilation asserts.
Use of dummy for some reason makes it work, otherwise offset is 8.
Static analysis in VS shows offset 4 in both cases.

Comment (x87) [reply cleo_plugins/Audio/CAudioStream.h:8]: hmm we need to figure out why it happens, otherwise there no guarantee this offset won't become 4 or 12 at some point

Comment (x87) [reply cleo_plugins/Audio/CAudioStream.h:8]: how can I trust this macro?

image

Comment (MiranDMC) [reply cleo_plugins/Audio/CAudioStream.h:8]: Macro works fine. Correct value is basing on CLEO4 is 4, we have 8 for some reason.
No clue why it happens.

Comment (x87) [reply cleo_plugins/Audio/CAudioStream.h:8]: I found this question https://stackoverflow.com/questions/27764464/why-is-vtable-pointer-aligned-to-8-bytes-on-x86-compiled-program

It links an article that seemed to have explained the compiler behavior in detail, but unfortunately it's unavailable and internet archive is not working at the moment. But there is an interesting bit:

> The 8 byte structure member causes the vfptr to be misaligned to 8 bytes rather than 4 bytes, as vfptr is added to structure after all the other members have been added/aligned. This behavior seems to be Visual Studio specific

It looks like addition of new `double` properties in CAudioStream changed default alignment from 4 to 8:

With `double`
image

With `float`
image

Even though the diagram shows `streamInternal` at offset +4 in both cases, it seems that compiler does some extra optimization and allocates it at offset +8 in the first case. 

Inheriting from an empty virtual class seems to actually help here and avoid this optimization, resulting in correct +4 offset. This trick should also be explained in that article. Here is a relevant discussion https://www.reddit.com/r/programming/comments/11ufrs/the_stolen_bytes_visual_studio_virtual_methods/


With that said, I think we can try the following:

1) make new properties (speed, volume, etc) `float` to enforce 4-byte alignment. I don't see why we need extra precision here.

2) If it's impossible (why?), then we can keep the dummy class.

Thoughts?

Comment (x87) [reply cleo_plugins/Audio/CAudioStream.h:8]: when IA is back online, we can try find [that article](http://lolengine.net/blog/2012/10/21/the-stolen-bytes) and read the explanation

Comment (x87) [reply cleo_plugins/Audio/CAudioStream.h:8]: and here is a reproduction using online c++ compiler 
https://cpp.godbolt.org/z/6d8qfqKfW

Comment (MiranDMC) [reply cleo_plugins/Audio/CAudioStream.h]: Yes, IA is currently under attack. How did you get these graphs of memory layout?
I changed types to float and it worked.

File: source/CleoBase.cpp [hunk 4]
```diff
@@ -252,9 +252,9 @@ namespace CLEO
         GetInstance().RemoveCallback(id, func);
     }
 
-    DWORD WINAPI CLEO_GetInternalAudioStream(CLEO::CRunningScript* thread, DWORD stream) // arg CAudioStream *
+    DWORD WINAPI CLEO_GetInternalAudioStream(CLEO::CRunningScript* unused, DWORD audioStreamPtr)
     {
-        return stream; // CAudioStream::streamInternal offset is 0
+        return audioStreamPtr + 0x4; // offsetof(CAudioStream, streamInternal)
```
Comment (x87) [inline source/CleoBase.cpp:257]: check what you return from here. I think you should dereference the pointer:

```
    DWORD WINAPI CLEO_GetInternalAudioStream(CLEO::CRunningScript* unused, DWORD audioStreamPtr)
    {
        return *(DWORD*)(audioStreamPtr + 0x4);
    }
```

File: cleo_plugins/Audio/CAudioStream.h (hunk 3 above)
Comment (x87) [reply cleo_plugins/Audio/CAudioStream.h]: https://devblogs.microsoft.com/visualstudio/size-alignment-and-memory-layout-insights-for-c-classes-structs-and-unions/#memory-layout

File: source/CleoBase.cpp (hunk 4 above)
Comment (MiranDMC) [reply source/CleoBase.cpp]: True. Fixed.

File: cleo_plugins/Audio/C3DAudioStream.cpp [hunk 5]
```diff
@@ -6,6 +6,8 @@ using namespace CLEO;
 
 C3DAudioStream::C3DAudioStream(const char* filepath) : CAudioStream()
 {
+    static_assert(offsetof(C3DAudioStream, streamInternal) == 4, "C3DAudioStream compatibility with CLEO4 broken!");
```
Comment (x87) [inline cleo_plugins/Audio/C3DAudioStream.cpp:9]:
```suggestion
    // see https://github.com/cleolibrary/CLEO5/pull/230
    static_assert(offsetof(C3DAudioStream, streamInternal) == 4 && alignof(C3DAudioStream) == 4, "C3DAudioStream compatibility with CLEO4 broken!");
```

File: cleo_plugins/Audio/CAudioStream.cpp [hunk 6]
```diff
@@ -6,6 +6,8 @@ using namespace CLEO;
 
 CAudioStream::CAudioStream(const char* filepath)
 {
+    static_assert(offsetof(CAudioStream, streamInternal) == 4, "CAudioStream compatibility with CLEO4 broken!");
```
Comment (x87) [inline cleo_plugins/Audio/CAudioStream.cpp:9]:
```suggestion
    // see https://github.com/cleolibrary/CLEO5/pull/230
    static_assert(offsetof(CAudioStream, streamInternal) == 4 && alignof(CAudioStream) == 4, "CAudioStream compatibility with CLEO4 broken!");
```

**Review by x87 — CHANGES_REQUESTED** [review]

File: cleo_plugins/Audio/C3DAudioStream.cpp (hunk 5 above)
Comment (MiranDMC) [reply cleo_plugins/Audio/C3DAudioStream.cpp:9]: Does `alignof(C3DAudioStream)` has any meaning in that case? All we care of is final offset of `streamInternal` in relation to instance pointer.

Comment (x87) [reply cleo_plugins/Audio/C3DAudioStream.cpp:9]: Yes it has meaning. If you see above discussion, even though `streamInternal` had offset +4 in IDE the final offset was either +4 or +8 depending on the alignment of the class. This is why we got rid of 8-byte long fields.

Comment (x87) [reply cleo_plugins/Audio/C3DAudioStream.cpp:9]: if you add `double` back, the `alignof` becomes 8 and assert fail, which is what we want.

Comment (MiranDMC) [reply cleo_plugins/Audio/C3DAudioStream.cpp:9]: 'offsetof' fails too, but during compilation time.

**Review by x87 — APPROVED** [review]

File: cleo_plugins/Audio/CAudioStream.h (hunk 3 above)
Comment (x87) [reply cleo_plugins/Audio/CAudioStream.h]: @MiranDMC IA is back online and the article is here
https://web.archive.org/web/20190115013504/http://lolengine.net/blog/2012/10/21/the-stolen-bytes

TLDR;
> The problem is that the compiler decides to add the vtable pointer after it has aligned the class data, resulting in excessive realignment.
